meson: don't add `-pthread` to static linking flags on Windows

bgilbert opened the pull request:
Meson always returns `-pthread` in `dependency('threads')` on non-MSVC compilers.  On Windows we use Windows threading primitives, so we don't need this.  Avoid adding `-pthread` to libzstd's link flags, either as a Meson subproject or via pkg-config `Libs.private`, so the application doesn't inadvertently depend on winpthreads.

Add a Meson MinGW cross-compile CI test that checks for this.  It turns out that `pzstd` fails to build in that environment, so have the test skip building `contrib` for now.

bgilbert committed:
meson: don't add -pthread to static linking flags on Windows

Meson always returns -pthread in dependency('threads') on non-MSVC
compilers.  On Windows we use Windows threading primitives, so we don't
need this.  Avoid adding -pthread to libzstd's link flags, either as a
Meson subproject or via pkg-config Libs.private, so the application
doesn't inadvertently depend on winpthreads.

Add a Meson MinGW cross-compile CI test that checks for this.  It turns
out that pzstd fails to build in that environment, so have the test
skip building contrib for now.

bgilbert commented:
Is there anything I can do to help this move forward?

Cyan4973 commented:
It's mostly a matter of finding someone for the review.

In my understanding, this PR is helpful when using `meson` to build `zstd` on Windows.

It seems this is a combination we don't have in CI, hence we can't "witness" the impact of this change.
An opened question at this stage is if it would be worthwhile to add (and if yes: how, and what's the runtime cost).

bgilbert commented:
It's helpful for building zstd for Windows with MinGW, possibly via cross-compilation.  Many application stacks will already use winpthreads, either because something in the stack uses pthreads explicitly, or because something uses C/C++ thread intrinsics and MinGW-w64 is configured to implement those via pthreads.  In that case, zstd's current behavior amounts to harmless overlinking.  My stack uses MinGW-w64 configured to implement intrinsics via Win32 primitives (a less common configuration), and does not otherwise use or ship winpthreads, so the current behavior produces an unsatisfied dependency.

If desired, I can see about adding a short test job for MinGW + Meson, or provide manual build instructions for verification.

Cyan4973 commented:
That would be a good idea.
It would help preserve this property across future modifications,
otherwise another contributor might come and change a detail which breaks this property, and we wouldn't notice.

bgilbert commented:
Added a CI job.

It turns out that the DLL doesn't end up linked against winpthreads, so the problem is limited to the linker flags we tell applications to use when linking with our static library.  I've updated the PR description and commit message accordingly.

Cyan4973 reviewed (approved):
This looks good to me!

Cyan4973 commented:
I assume the last remaining CI error (`cygwin`) is unrelated

bgilbert commented:
Yes, looks like the Cygwin failure is a transient network issue.

bgilbert commented:
Is this ready to go in?

Cyan4973 commented:
> Is this ready to go in?

yep
